modules/services/payment/src/main/kotlin/au/com/shiftyjelly/pocketcasts/payment/FakePaymentDataSource.kt

```
@@ -10,48 +10,65 @@ import com.android.billingclient.api.QueryPurchaseHistoryParams
 
     override suspend fun loadProducts(): PaymentResult<List<Product>> {
-        return customProductsResult ?: PaymentResult.Success(KnownProducts)
+        return if (loadedProductsResultCode == PaymentResultCode.Ok) {
+            PaymentResult.Success(loadedProducts)
+        } else {
+            PaymentResult.Failure(loadedProductsResultCode, "Error message")
+        }
     }
 
     override suspend fun loadPurchases(): PaymentResult<List<Purchase>> {
-        return customPurchasesResult ?: PaymentResult.Success(DefaultPurchases)
-    }
-
-    override suspend fun acknowledgePurchase(purchase: Purchase): PaymentResult<Purchase> {
-        receivedPurchases += purchase
-        val resultCode = acknowledgeFlow.first()
-        return if (resultCode is PaymentResultCode.Ok) {
-            PaymentResult.Success(purchase.copy(isAcknowledged = true))
+        return if (loadedPurchasesResultCode == PaymentResultCode.Ok) {
+            PaymentResult.Success(loadedPurchases)
         } else {
-            PaymentResult.Failure(resultCode, "Error")
+            PaymentResult.Failure(loadedPurchasesResultCode, "Error message")
         }
     }
 
-    suspend fun emitAcknowledgeResponse(code: PaymentResultCode) {
-        acknowledgeFlow.emit(code)
+    override suspend fun launchBillingFlow(key: SubscriptionPlan.Key, activity: Activity): PaymentResult<Unit> {
+        return if (billingFlowResultCode == PaymentResultCode.Ok) {
+            val billingResult = PaymentResult.Success(Unit)
+            val purchaseResult = if (purchasedProductsResultCode == PaymentResultCode.Ok) {
+                PaymentResult.Success(purchasedProducts)
+            } else {
+                PaymentResult.Failure(purchasedProductsResultCode, "Error message")
+            }
+            _purchaseResults.emit(purchaseResult)
+            billingResult
+        } else {
+            PaymentResult.Failure(billingFlowResultCode, "Error message")
```

⛏️ I know it's just the fake data source, but maybe we could add a more describing error message to distinguish this case from the other `Failure` case above